### PR TITLE
draft: add TYPE command

### PR DESCRIPTION
**Pull request by tbmreza**
How do you test a new command? I'm not sure if I'm already on the right direction, i.e., if I only need to figure out what to feed `send_packed_command()` with, and how to receive the connection's response.

#28

### CODE REVIEW

**Comment by clarity0** [comment]
@tbmreza I havent tested it out extensively but according to the [redis-rs docs](https://docs.rs/redis/0.20.0/redis/struct.Connection.html#method.send_packed_command) changes the state of the connection or something, and my test for the GETDEL command only behaved as expected when I called get twice. I just gave up testing that out because no Redis client I've seen implements GETDEL and it's deprecated apparently. [see my comment here](https://github.com/Qovery/RedisLess/pull/29#discussion_r628715265)

**Comment by clarity0** [comment]
I think the [python redis client](https://redis-py.readthedocs.io/en/stable/index.html?highlight=type#redis.Redis.type) does implement TYPE unlike redis-rs. You could test it out with that.

File: redisless/storage/src/in_memory.rs [hunk 1]
```diff
@@ -113,6 +113,19 @@ impl Storage for InMemoryStorage {
...
+    fn value_type(&mut self, key: &[u8]) -> String {
+        if let Some(data_type) = self.data_mapper.get(key) {
+            match data_type {
+                DataType::String => format!("string"),
+                DataType::List => format!("list"),
+                DataType::Set => format!("set"),
+                DataType::Hash => format!("hash"),
+            }
+        } else {
+            format!("none")
```
[the remark this answers is not in the document]
Comment (clarity0) [reply redisless/storage/src/in_memory.rs:126]: Also you can match directly on the Some inner type like `Some(DataType::String) => ...` , that way you can match directly on the datamapper.get and avoid nesting it like that with an if let. Also I think it's better if you return a `&str` from value_type, that way you don't need to allocate a whole `String`

[the remark this answers is not in the document]
Comment (tbmreza) [reply redisless/storage/src/in_memory.rs]: @clarity0 Got it. I have just:
- changed the return type to &str
- matched directly without if let nest
- added the command in the python client

Is that everything?

**Comment by clarity0** [comment]
Does it work well with the python client?

**Comment by tbmreza** [comment]
> Does it work well with the python client?

Should I test it thoroughly? If so, do I put them (SET and TYPE commands for every DataType) all in main.py? @clarity0

**Comment by tbmreza** [comment]
for the 1 TYPE command I tested in python client, yes it works well (the assertion doesn't fail)

**Comment by clarity0** [comment]
Right now there is only support commands that add strings so you can't really test the other cases with a client but they should work fine.

**Comment by evoxmusic** [comment]
@clarity0 @tbmreza good to be merged?

**Comment by evoxmusic** [comment]
@tbmreza can you rebase?

**Comment by tbmreza** [comment]
> @tbmreza can you rebase?

sure but I'm not a capable git user. could you tell me how to do it, step by step?

**Comment by tbmreza** [comment]
I'm closing this PR. Currently looking at other good first issues (#23 if it doesn't turn out to be too hard for me), will open new PR covering that and #28. Thanks for bearing with me!

**Pull request closed without merging**